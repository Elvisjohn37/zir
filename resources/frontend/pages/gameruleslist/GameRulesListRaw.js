import React from 'react';
import Container from '@mui/material/Container';
import Grid from '@mui/material/Grid';
import Button from '@mui/material/Button';
import styles from './GameRulesListRaw.module';
import InsertDriveFileIcon from '@mui/icons-material/InsertDriveFile';
import classnames from 'classnames';
import IconButton from '@mui/material/IconButton';
import ArrowBackIcon from '@mui/icons-material/ArrowBack';
import { useLocation, NavLink, useParams } from 'react-router-dom';
import lodash from 'lodash';
import GameRulesLoader from 'frontend/components/GameRulesLoader';
import GameRulesErrorMessage from 'frontend/components/GameRulesErrorMessage';
import { FormattedMessage } from 'react-intl';

export default function GameRulesListRaw({ gameRulesState }) {
	let location = useLocation();
	let { product, gameType } = useParams();
	let currentPath = location.pathname;
	let splitPathName = currentPath.split('/');
	splitPathName.pop();
	splitPathName.pop();
	let rootPath = splitPathName.join('/');
	let filteredGameRules = gameRulesState.gameRulesList.filter(
		(item) => lodash.kebabCase(item.productName) == product && lodash.kebabCase(item.gameTypeName) == gameType
	);

	return (
		<Container className={styles.gameRulesLisRaw}>
			<div className={styles.backButtonContainer}>
				<IconButton component={NavLink} to={rootPath}>
					<ArrowBackIcon />
				</IconButton>
			</div>
			<Grid container spacing={1}>
				{gameRulesState.isLoading ? (
					<GameRulesLoader />
				) : gameRulesState.isError || filteredGameRules.length == 0 ? (
					<GameRulesErrorMessage />
				) : (
					filteredGameRules.map((item, index) => {
						return (
							<Grid item md={6} xs={12} lg={4} key={index}>
								<Button
									component={NavLink}
									startIcon={<InsertDriveFileIcon />}
									variant="contained"
									disableElevation
									color="primary"
									className={classnames([styles.button, item.isNew == '1' && styles.isNew])}
									to={`${currentPath}/${lodash.kebabCase(item.gameName)}`}
								>
									{item.gameName}
									{item.isNew == '1' && (
										<span className={styles.newTag}>
											<FormattedMessage id="new" />
										</span>
									)}
								</Button>
							</Grid>
						);
					})
				)}
			</Grid>
		</Container>
	);
}
