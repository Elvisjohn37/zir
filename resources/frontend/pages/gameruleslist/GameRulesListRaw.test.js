import React from 'react';
import GameRulesListRaw from './GameRulesListRaw';
import { render } from '@testing-library/react';
import * as gameRulesHelper from 'frontend/tests/gameRulesHelper';
import * as reactRouterDom from 'react-router-dom';
import * as lodash from 'lodash';

jest.mock('@mui/material/Container');
jest.mock('@mui/material/Grid');
jest.mock('@mui/material/Button');
jest.mock('@mui/icons-material/InsertDriveFile');
jest.mock('classnames');
jest.mock('@mui/material/IconButton');
jest.mock('@mui/icons-material/ArrowBack');
jest.mock('react-router-dom');
jest.mock('lodash');
jest.mock('frontend/components/GameRulesLoader');
jest.mock('frontend/components/GameRulesErrorMessage');

describe('<GameRulesListRaw/>', () => {
	beforeEach(() => {
		gameRulesHelper.resetGameRulesState();
		reactRouterDom.resetUseParams();
	});

	it('should match snapshot', () => {
		reactRouterDom.setUseParams({ product: 'games', gameType: 'table-games' });
		let { asFragment } = render(<GameRulesListRaw gameRulesState={gameRulesHelper.getGameRulesState()} />);
		expect(asFragment()).toMatchSnapshot();
	});

	it('should display loader while game rules list data is loading', () => {
		reactRouterDom.setUseParams({ product: 'games', gameType: 'table-games' });

		gameRulesHelper.setGameRulesState({
			isLoading: true,
		});

		let { getByTestId, queryByTestId, rerender } = render(
			<GameRulesListRaw gameRulesState={gameRulesHelper.getGameRulesState()} />
		);

		expect(getByTestId('GameRulesLoader')).toBeInTheDocument();

		gameRulesHelper.setGameRulesState({
			isLoading: false,
		});

		rerender(<GameRulesListRaw gameRulesState={gameRulesHelper.getGameRulesState()} />);

		expect(queryByTestId('GameRulesLoader')).not.toBeInTheDocument();
	});

	it('should display game rules list correctly', () => {
		reactRouterDom.setUseParams({ product: 'games', gameType: 'table-games' });
		let currentPath = reactRouterDom.useLocation().pathname;
		let gameRulesState = gameRulesHelper.getGameRulesState();
		let { baseElement, rerender } = render(
			<GameRulesListRaw gameRulesState={gameRulesHelper.getGameRulesState()} />
		);
		let filteredGameRules = gameRulesState.gameRulesList.filter(
			(item) =>
				lodash.kebabCase(item.productName) == 'games' && lodash.kebabCase(item.gameTypeName) == 'table-games'
		);
		let gameRulesListDom = baseElement.querySelectorAll('.button');

		filteredGameRules.map((gr, index) => {
			let gameRulesListItem = gameRulesListDom[index];
			let isNewtag = gameRulesListItem.querySelector('.newTag');

			expect(gameRulesListItem).toHaveAttribute('to', `${currentPath}/${lodash.kebabCase(gr.gameName)}`);

			if (gr.isNew) {
				expect(gameRulesListItem).toHaveClass('isNew');
				expect(isNewtag).toBeInTheDocument();
			} else {
				expect(gameRulesListItem).not.toHaveClass('isNew');
				expect(isNewtag).not.toBeInTheDocument();
			}
		});

		reactRouterDom.setUseParams({ product: 'multiplayer', gameType: 'card-games' });
		rerender(<GameRulesListRaw gameRulesState={gameRulesHelper.getGameRulesState()} />);
		filteredGameRules = gameRulesState.gameRulesList.filter(
			(item) =>
				lodash.kebabCase(item.productName) == 'multiplayer' &&
				lodash.kebabCase(item.gameTypeName) == 'card-games'
		);
		gameRulesListDom = baseElement.querySelectorAll('.button');

		filteredGameRules.map((gr, index) => {
			let gameRulesListItem = gameRulesListDom[index];
			let isNewtag = gameRulesListItem.querySelector('.newTag');

			expect(gameRulesListItem).toHaveAttribute('to', `${currentPath}/${lodash.kebabCase(gr.gameName)}`);

			if (gr.isNew) {
				expect(gameRulesListItem).toHaveClass('isNew');
				expect(isNewtag).toBeInTheDocument();
			} else {
				expect(gameRulesListItem).not.toHaveClass('isNew');
				expect(isNewtag).not.toBeInTheDocument();
			}
		});
	});

	it('should display error message when game rules list data loading failed', () => {
		reactRouterDom.setUseParams({ product: 'multiplayer', gameType: 'card-games' });

		let { getByTestId, queryByTestId, rerender } = render(
			<GameRulesListRaw gameRulesState={gameRulesHelper.getGameRulesState()} />
		);

		expect(queryByTestId('GameRulesErrorMessage')).not.toBeInTheDocument();

		gameRulesHelper.setGameRulesState({
			isError: true,
		});

		rerender(<GameRulesListRaw gameRulesState={gameRulesHelper.getGameRulesState()} />);

		expect(getByTestId('GameRulesErrorMessage')).toBeInTheDocument();
	});
});
