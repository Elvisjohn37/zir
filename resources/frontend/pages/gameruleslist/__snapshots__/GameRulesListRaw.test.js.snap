// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameRulesListRaw/> should match snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="Container"
  >
    <div
      class="backButtonContainer"
    >
      <button
        data-testid="IconButton"
        to=""
      >
        <div
          data-testid="ArrowBack"
        />
        <div>
          {}
        </div>
      </button>
    </div>
    <div
      container="true"
      data-testid="Grid"
      spacing="1"
    >
      <div
        data-testid="Grid"
        item="true"
        lg="4"
        md="6"
        xs="12"
      >
        <a
          class="button"
          color="primary"
          data-testid="NavLink"
          disableelevation="true"
          to="www.zirconloc.com/path/dice-wars-mobile"
          variant="contained"
        >
          Dice Wars (Mobile)
          <div>
            {}
          </div>
        </a>
      </div>
      <div
        data-testid="Grid"
        item="true"
        lg="4"
        md="6"
        xs="12"
      >
        <a
          class="button isNew"
          color="primary"
          data-testid="NavLink"
          disableelevation="true"
          to="www.zirconloc.com/path/djap-go-mobile"
          variant="contained"
        >
          Djap Go (Mobile)
          <span
            class="newTag"
          >
            <div
              data-testid="FormattedMessage"
              id="new"
            >
              <div>
                new
              </div>
            </div>
          </span>
          <div>
            {}
          </div>
        </a>
      </div>
      <div
        data-testid="Grid"
        item="true"
        lg="4"
        md="6"
        xs="12"
      >
        <a
          class="button"
          color="primary"
          data-testid="NavLink"
          disableelevation="true"
          to="www.zirconloc.com/path/sedie-mobile"
          variant="contained"
        >
          Sedie (Mobile)
          <div>
            {}
          </div>
        </a>
      </div>
      <div
        data-testid="Grid"
        item="true"
        lg="4"
        md="6"
        xs="12"
      >
        <a
          class="button isNew"
          color="primary"
          data-testid="NavLink"
          disableelevation="true"
          to="www.zirconloc.com/path/three-stars-blessing-mobile"
          variant="contained"
        >
          Three Stars Blessing (Mobile)
          <span
            class="newTag"
          >
            <div
              data-testid="FormattedMessage"
              id="new"
            >
              <div>
                new
              </div>
            </div>
          </span>
          <div>
            {}
          </div>
        </a>
      </div>
      <div
        data-testid="Grid"
        item="true"
        lg="4"
        md="6"
        xs="12"
      >
        <a
          class="button"
          color="primary"
          data-testid="NavLink"
          disableelevation="true"
          to="www.zirconloc.com/path/whack-d-mole-mobile"
          variant="contained"
        >
          Whack d' Mole (Mobile)
          <div>
            {}
          </div>
        </a>
      </div>
      <div
        data-testid="Grid"
        item="true"
        lg="4"
        md="6"
        xs="12"
      >
        <a
          class="button"
          color="primary"
          data-testid="NavLink"
          disableelevation="true"
          to="www.zirconloc.com/path/yin-yang-treasure-mobile"
          variant="contained"
        >
          Yin Yang Treasure (Mobile)
          <div>
            {}
          </div>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
